Refactor: Workspace Factory Generator (1/1)

chynu opened the pull request:
Refactored 13/13 functions from `wfactory_generator.js` into new refactor files.

Some of the functions here will be deleted in future PR's because it no longer fits with our design of DevTools (e.g. importing defined blocks should be done only through the block library editor, since any blocks in a toolbox should be defined in a library in that same project anyway).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/77)
<!-- Reviewable:end -->

chynu committed:
Moved functions from wfactory_generator.js

chynu committed:
Merge branch 'develop' into rf_wsGenerator

chynu committed:
Moved more functions from generator

chynu committed:
Merge branch 'develop' into rf_wsGenerator

chynu committed:
Moved last few fcn's from generator

chynu committed:
a singular nit

chynu committed:
function rename

svouse commented:
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/model/project.js, line 338 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoK9fVbaMeRSwh-UUrL:-KoK9fVbaMeRSwh-UUrM:b-5ez23c) ([raw file](https://github.com/google/blockly-devtools/blob/2ce2a3a6251614dc20d21cab88f4c9b27bffb8bb/src/model/project.js#L338)):*
> ```JavaScript
>    */
>   generateInjectString() {
>     /*
> ```

I'm wondering if this sort of functionality belongs in a controller. Can you explain why it works best in project//its ultimate place in the refactor? 

---

*[src/model/toolbox.js, line 109 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKAsN9z9l23yiSzUJw:-KoKAsN9z9l23yiSzUJx:bb3xe6l) ([raw file](https://github.com/google/blockly-devtools/blob/2ce2a3a6251614dc20d21cab88f4c9b27bffb8bb/src/model/toolbox.js#L109)):*
> ```JavaScript
>      */
>     console.log('generateXml() called.');
>   }
> ```

Largely just a clarifying question//more post-refactor-relevant; for export of model elements, do we intend to keep these functions? I suppose I'm wondering whether we'll eventually just have getExportData or if I should get rid of it/we still need multiple functions

---

*[src/model/workspace_contents.js, line 160 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKBN3hyLDoSPW73C4b:-KoKBN3hyLDoSPW73C4c:b-ya81ng) ([raw file](https://github.com/google/blockly-devtools/blob/2ce2a3a6251614dc20d21cab88f4c9b27bffb8bb/src/model/workspace_contents.js#L160)):*
> ```JavaScript
>      */
>     console.log('generateJsFileContents() called from WorkspaceContents.');
>   }
> ```

Same as above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/77)*
<!-- Sent from Reviewable.io -->

chynu committed:
Moved functions (#77)

chynu commented:
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/model/toolbox.js, line 109 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKAsN9z9l23yiSzUJw:-KoKDNTIGjuf0YGxI7EL:b-mnhg5k) ([raw file](https://github.com/google/blockly-devtools/blob/2ce2a3a6251614dc20d21cab88f4c9b27bffb8bb/src/model/toolbox.js#L109)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Largely just a clarifying question//more post-refactor-relevant; for export of model elements, do we intend to keep these functions? I suppose I'm wondering whether we'll eventually just have getExportData or if I should get rid of it/we still need multiple functions
</blockquote></details>

We need both functions. `generateXml()`(or now `getExportData()`) for just the XML and `generateJsFileContents()` for generating the other contents in the exported file.

As per our discussion, moved `generateJsFileContents()` to the controller level.

---

*[src/model/project.js, line 338 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoK9fVbaMeRSwh-UUrL:-KoKC5d6OfK3PsLw0WDA:b-8oxxa) ([raw file](https://github.com/google/blockly-devtools/blob/2ce2a3a6251614dc20d21cab88f4c9b27bffb8bb/src/model/project.js#L338)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

I'm wondering if this sort of functionality belongs in a controller. Can you explain why it works best in project//its ultimate place in the refactor? 
</blockquote></details>

Oops you're right ;~; Moved to `project_controller.js`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/77)*
<!-- Sent from Reviewable.io -->

svouse commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/77#-:-KoKQdltEoPXgVw7o7__:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 3 of 8 files at r1, 4 of 6 files at r2.
Review status: 7 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 420 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKQnqHWygOJ1DARcdk:-KoKQnqHWygOJ1DARcdl:bbjp655) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/controller/toolbox_controller.js#L420)):*
> ```JavaScript
>    */
>   extractToolboxXml() {
>     let extractedToolbox = new Object();
> ```

Why `new Object()` instead of `{}`?

---

*[src/controller/workspace_controller.js, line 243 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKQw7Cl5lF3e7gPCYJ:-KoKQw7Cl5lF3e7gPCYK:b-n545af) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/controller/workspace_controller.js#L243)):*
> ```JavaScript
>    */
>   extractWorkspaceContentsXml() {
>     let extractedWorkspaceContents = new Object();
> ```

Same question here.

---

*[src/model/toolbox.js, line 72 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKR5uscbeWfjPE81S5:-KoKR5uscbeWfjPE81S6:b-n7l345) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/model/toolbox.js#L72)):*
> ```JavaScript
> 
>   /**
>    * Generates XML string representation of Toolbox for user to download. Does
> ```

Is this still guaranteed to be XML?

---

*[src/model/workspace_contents.js, line 118 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKRFpIk68817CcNg0Z:-KoKRFpIk68817CcNg0_:b-r26nxz) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/model/workspace_contents.js#L118)):*
> ```JavaScript
> 
>   /**
>    * Generates string of XML dom element for WorkspaceContents. Used to insert
> ```

DOM is capitalized.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/77)*
<!-- Sent from Reviewable.io -->

chynu committed:
Minor changes for #77

chynu commented:
Review status: 7 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 420 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKQnqHWygOJ1DARcdk:-KoKb4erxtAs7czsvCaf:b-9h4ee5) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/controller/toolbox_controller.js#L420)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Why `new Object()` instead of `{}`?
</blockquote></details>

Was mostly an arbitrary decision. Is there a stylistic preference/advantage to using `{}`?

---

*[src/controller/workspace_controller.js, line 243 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKQw7Cl5lF3e7gPCYJ:-KoKcCtf0EZVxFdJlOzR:b-1ade80) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/controller/workspace_controller.js#L243)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Same question here.
</blockquote></details>

Is there a preference for one over the other?

---

*[src/model/toolbox.js, line 72 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKR5uscbeWfjPE81S5:-KoKcFm-stGljqc4QY78:b-4536dz) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/model/toolbox.js#L72)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Is this still guaranteed to be XML?
</blockquote></details>

I changed return type to XML DOM elements (here and in `WorkspaceContents`). Probably best that it is converted to string if necessary at controller level (?).

---

*[src/model/workspace_contents.js, line 118 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKRFpIk68817CcNg0Z:-KoKcM-W34wApP78oYk5:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/model/workspace_contents.js#L118)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

DOM is capitalized.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/77)*
<!-- Sent from Reviewable.io -->

chynu committed:
Merge branch 'develop' into rf_wsGenerator

AnmAtAnm commented on src/controller/workspace_controller.js:
```diff
@@ -232,4 +232,44 @@ class WorkspaceController {
+   * Generates JavaScript string representation of WorkspaceContents for user to
+   * download. Does not deal with popups or file system access; just generates content.
+   *
+   * @returns {string} String representation of JS file to be exported.
```
Throw "unimplemented" if you are not returning a stand-in string.

chynu replied:
Added `throw 'Unimplemented: functionName()'` throughout PR.

AnmAtAnm commented on src/controller/toolbox_controller.js:
```diff
@@ -411,6 +411,46 @@ class ToolboxController {
+   * Generates JavaScript string representation of toolbox for user to download.
+   * Does not deal with popups or file system access; just generates content.
+   *
+   * @returns {string} String representation of JS file to be exported.
```
Throw "unimplemented" if you are not returning a stand-in string.

AnmAtAnm commented on src/controller/toolbox_controller.js:
```diff
@@ -411,6 +411,46 @@ class ToolboxController {
   }
 
   /**
+   * Extracts Toolbox XML from user-uploaded file. XML should be between marked
```
Is there an input parameter from which this is extracted?

chynu replied:
Added a string parameter.

AnmAtAnm commented on src/controller/project_controller.js:
```diff
@@ -223,6 +223,22 @@ class ProjectController {
+   * a Blockly application.
+   * Auto-injects added toolbox into the inject function.
+   *
+   * @return {string} String representation of starter code for inject file.
```
Throw "unimplemented" since this does not yet return anything.

AnmAtAnm commented on src/controller/editor_controller.js:
```diff
@@ -109,4 +109,74 @@ class EditorController {
+   * type.
+   * @param {!Array.<!Element>} blockTypes Array of blocks that have been defined.
+   * @return {!Array.<!Blockly.Block>} Array of Blockly.Block objects corresponding
+   *     to the array of blockTypes.
```
Throw "unimplemented" since this does not yet return anything.

AnmAtAnm commented on src/controller/toolbox_controller.js:
```diff
@@ -411,6 +411,46 @@ class ToolboxController {
+     * References:
+     * - evaluateMarkedCode()
+     */
+
```
Console log unimplemented.

AnmAtAnm commented on src/controller/workspace_controller.js:
```diff
@@ -232,4 +232,44 @@ class WorkspaceController {
+
+  /**
+   * Extracts WorkspaceContents XML from user-uploaded file. XML should be between
+   * marked comments in uploaded file.
```
Where is this user uploaded file?  I expected an input parameter.

chynu replied:
Added a parameter.

AnmAtAnm commented on src/view/toolbox_editor_view.js:
```diff
@@ -239,4 +239,22 @@ class ToolboxEditorView {
+     * References:
+     * - evaluateMarkedCode()
+     */
+  }
```
Console.log "unimplemented ToolboxEditorView.loadImportedToolbox()"

AnmAtAnm commented on src/view/workspace_editor_view.js:
```diff
@@ -73,5 +73,21 @@ class WorkspaceEditorView {
+     * References:
+     * - evaluateMarkedCode()
+     */
+  }
```
Console.log "unimplemented WorkspaceEditorView.loadImportedWorkspaceContents()"

AnmAtAnm reviewed (commented):
Mostly good. I just want to track unimplemented methods with console log or throws clauses.

rachel-fenichel commented:
Review status: 5 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 420 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/77#-KoKQnqHWygOJ1DARcdk:-KoOALCGftpfcSM-2IKS:b3xjlsk) ([raw file](https://github.com/google/blockly-devtools/blob/13c33cc5339332cda89878f748917de457a8e7df/src/controller/toolbox_controller.js#L420)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Was mostly an arbitrary decision. Is there a stylistic preference/advantage to using `{}`?
</blockquote></details>

I have a slight stylistic preference toward `{}`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/77)*
<!-- Sent from Reviewable.io -->

chynu committed:
Added throws for unimplemented functions

chynu committed:
Merge branch 'develop' into rf_wsGenerator

chynu committed:
Adding more unimplemented errors

chynu committed:
Minor changes

AnmAtAnm reviewed (approved):
LGTM

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/77#-:-KoOSfQ8y4W0WW_GhR0s:bnfp4nl)*
<!-- Sent from Reviewable.io -->
